// @speckle/shared/dist/commonjs/workspaces/helpers/features.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspacePlanConfigs = exports.WorkspaceUnpaidPlanConfigs = exports.WorkspacePaidPlanConfigs = exports.WorkspacePlanFeaturesMetadata = exports.WorkspacePlanFeatures = void 0;
const plans_js_1 = require("./plans.js");
/**
 * WORKSPACE FEATURES
 */
exports.WorkspacePlanFeatures = {
    // Core features pretty much available to everyone
    Workspace: 'workspace',
    RoleManagement: 'roleManagement',
    GuestUsers: 'guestUsers',
    PrivateAutomateFunctions: 'privateAutomateFunctions',
    // Optional/plan specific
    DomainSecurity: 'domainBasedSecurityPolicies',
    PrioritySupport: 'prioritySupport',
    SSO: 'oidcSso',
    CustomDataRegion: 'workspaceDataRegionSpecificity'
};
exports.WorkspacePlanFeaturesMetadata = {
    // Old
    [exports.WorkspacePlanFeatures.Workspace]: {
        displayName: 'Workspace',
        description: 'A shared space for your team and projects'
    },
    [exports.WorkspacePlanFeatures.RoleManagement]: {
        displayName: 'Role management',
        description: "Control individual members' access and edit rights"
    },
    [exports.WorkspacePlanFeatures.GuestUsers]: {
        displayName: 'Guest users',
        description: (params) => `Give guests access to specific projects in the workspace at ${params.price}/month/guest`
    },
    [exports.WorkspacePlanFeatures.PrivateAutomateFunctions]: {
        displayName: 'Private automate functions',
        description: 'Create and manage private automation functions securely within your workspace'
    },
    [exports.WorkspacePlanFeatures.DomainSecurity]: {
        displayName: 'Domain security',
        description: 'Require workspace members to use a verified company email'
    },
    [exports.WorkspacePlanFeatures.SSO]: {
        displayName: 'Single Sign-On (SSO)',
        description: 'Require workspace members to log in with your SSO provider'
    },
    [exports.WorkspacePlanFeatures.CustomDataRegion]: {
        displayName: 'Custom data residency',
        description: 'Store the workspace data in a custom region'
    },
    [exports.WorkspacePlanFeatures.PrioritySupport]: {
        displayName: 'Priority support',
        description: 'Personal and fast support'
    }
};
const baseFeatures = [
    exports.WorkspacePlanFeatures.Workspace,
    exports.WorkspacePlanFeatures.RoleManagement,
    exports.WorkspacePlanFeatures.GuestUsers,
    exports.WorkspacePlanFeatures.PrivateAutomateFunctions
];
exports.WorkspacePaidPlanConfigs = {
    // Old
    [plans_js_1.PaidWorkspacePlans.Starter]: {
        plan: plans_js_1.PaidWorkspacePlans.Starter,
        features: [...baseFeatures, exports.WorkspacePlanFeatures.DomainSecurity]
    },
    [plans_js_1.PaidWorkspacePlans.Plus]: {
        plan: plans_js_1.PaidWorkspacePlans.Plus,
        features: [
            ...baseFeatures,
            exports.WorkspacePlanFeatures.DomainSecurity,
            exports.WorkspacePlanFeatures.SSO
        ]
    },
    [plans_js_1.PaidWorkspacePlans.Business]: {
        plan: plans_js_1.PaidWorkspacePlans.Business,
        features: [
            ...baseFeatures,
            exports.WorkspacePlanFeatures.DomainSecurity,
            exports.WorkspacePlanFeatures.SSO,
            exports.WorkspacePlanFeatures.CustomDataRegion,
            exports.WorkspacePlanFeatures.PrioritySupport
        ]
    },
    [plans_js_1.PaidWorkspacePlans.Team]: {
        plan: plans_js_1.PaidWorkspacePlans.Team,
        features: baseFeatures
    },
    [plans_js_1.PaidWorkspacePlans.Pro]: {
        plan: plans_js_1.PaidWorkspacePlans.Pro,
        features: [
            ...baseFeatures,
            exports.WorkspacePlanFeatures.DomainSecurity,
            exports.WorkspacePlanFeatures.SSO,
            exports.WorkspacePlanFeatures.CustomDataRegion,
            exports.WorkspacePlanFeatures.PrioritySupport
        ]
    }
};
exports.WorkspaceUnpaidPlanConfigs = {
    // Old
    [plans_js_1.UnpaidWorkspacePlans.Unlimited]: {
        plan: plans_js_1.UnpaidWorkspacePlans.Unlimited,
        features: [
            ...baseFeatures,
            exports.WorkspacePlanFeatures.DomainSecurity,
            exports.WorkspacePlanFeatures.SSO,
            exports.WorkspacePlanFeatures.CustomDataRegion,
            exports.WorkspacePlanFeatures.PrioritySupport
        ]
    },
    [plans_js_1.UnpaidWorkspacePlans.Academia]: {
        plan: plans_js_1.UnpaidWorkspacePlans.Academia,
        features: [
            ...baseFeatures,
            exports.WorkspacePlanFeatures.DomainSecurity,
            exports.WorkspacePlanFeatures.SSO,
            exports.WorkspacePlanFeatures.CustomDataRegion,
            exports.WorkspacePlanFeatures.PrioritySupport
        ]
    },
    [plans_js_1.UnpaidWorkspacePlans.StarterInvoiced]: {
        ...exports.WorkspacePaidPlanConfigs.starter,
        plan: plans_js_1.UnpaidWorkspacePlans.StarterInvoiced
    },
    [plans_js_1.UnpaidWorkspacePlans.PlusInvoiced]: {
        ...exports.WorkspacePaidPlanConfigs.plus,
        plan: plans_js_1.UnpaidWorkspacePlans.PlusInvoiced
    },
    [plans_js_1.UnpaidWorkspacePlans.BusinessInvoiced]: {
        ...exports.WorkspacePaidPlanConfigs.business,
        plan: plans_js_1.UnpaidWorkspacePlans.BusinessInvoiced
    },
    // New
    [plans_js_1.UnpaidWorkspacePlans.Free]: {
        plan: plans_js_1.UnpaidWorkspacePlans.Free,
        features: baseFeatures
    }
};
exports.WorkspacePlanConfigs = {
    ...exports.WorkspacePaidPlanConfigs,
    ...exports.WorkspaceUnpaidPlanConfigs
};
