// @speckle/shared/dist/commonjs/workspaces/helpers/plans.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspacePlanStatuses = exports.TrialWorkspacePlanStatuses = exports.PaidWorkspacePlanStatuses = exports.UnpaidWorkspacePlanStatuses = exports.WorkspacePlanBillingIntervals = exports.isNewWorkspacePlan = exports.WorkspaceGuestSeatType = exports.WorkspacePlans = exports.UnpaidWorkspacePlans = exports.PaidWorkspacePlans = exports.PaidWorkspacePlansNew = exports.PaidWorkspacePlansOld = exports.TrialEnabledPaidWorkspacePlans = void 0;
/**
 * PLANS
 */
exports.TrialEnabledPaidWorkspacePlans = {
    Starter: 'starter'
};
exports.PaidWorkspacePlansOld = {
    ...exports.TrialEnabledPaidWorkspacePlans,
    Plus: 'plus',
    Business: 'business'
};
exports.PaidWorkspacePlansNew = {
    Team: 'team',
    Pro: 'pro'
};
exports.PaidWorkspacePlans = {
    ...exports.PaidWorkspacePlansOld,
    ...exports.PaidWorkspacePlansNew
};
exports.UnpaidWorkspacePlans = {
    // Old
    StarterInvoiced: 'starterInvoiced',
    PlusInvoiced: 'plusInvoiced',
    BusinessInvoiced: 'businessInvoiced',
    // New
    Unlimited: 'unlimited',
    Academia: 'academia',
    Free: 'free'
};
exports.WorkspacePlans = {
    ...exports.PaidWorkspacePlans,
    ...exports.UnpaidWorkspacePlans
};
// TODO: Remove this post workspace migration
exports.WorkspaceGuestSeatType = 'guest';
// TODO: Remove this post workspace migration, only needed temporarily to differiante between old and new
const isNewWorkspacePlan = (plan) => {
    return (plan === exports.PaidWorkspacePlansNew.Team ||
        plan === exports.PaidWorkspacePlansNew.Pro ||
        plan === exports.UnpaidWorkspacePlans.Free);
};
exports.isNewWorkspacePlan = isNewWorkspacePlan;
/**
 * BILLING INTERVALS
 */
exports.WorkspacePlanBillingIntervals = {
    Monthly: 'monthly',
    Yearly: 'yearly'
};
/**
 * PLAN STATUSES
 */
exports.UnpaidWorkspacePlanStatuses = {
    Valid: 'valid'
};
exports.PaidWorkspacePlanStatuses = {
    ...exports.UnpaidWorkspacePlanStatuses,
    PaymentFailed: 'paymentFailed',
    CancelationScheduled: 'cancelationScheduled',
    Canceled: 'canceled'
};
exports.TrialWorkspacePlanStatuses = {
    Trial: 'trial',
    Expired: 'expired'
};
exports.WorkspacePlanStatuses = {
    ...exports.PaidWorkspacePlanStatuses,
    ...exports.TrialWorkspacePlanStatuses,
    ...exports.UnpaidWorkspacePlanStatuses
};
