// @speckle/shared/dist/commonjs/authz/domain/authErrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceSsoSessionInvalidError = exports.WorkspaceNoAccessError = exports.ProjectNoAccessError = exports.ProjectNotFoundError = exports.defineAuthError = void 0;
const defineAuthError = (params) => {
    const { code, message } = params;
    return {
        code,
        message
    };
};
exports.defineAuthError = defineAuthError;
exports.ProjectNotFoundError = (0, exports.defineAuthError)({
    code: 'ProjectNotFound',
    message: 'Project not found'
});
exports.ProjectNoAccessError = (0, exports.defineAuthError)({
    code: 'ProjectNoAccess',
    message: 'You do not have access to the project'
});
exports.WorkspaceNoAccessError = (0, exports.defineAuthError)({
    code: 'WorkspaceNoAccess',
    message: 'You do not have access to the workspace'
});
exports.WorkspaceSsoSessionInvalidError = (0, exports.defineAuthError)({
    code: 'WorkspaceSsoSessionInvalid',
    message: 'Your workspace SSO session is invalid'
});
